Userland/Libraries/LibGfx/ImageFormats/JPEGWriter.cpp

```diff
@@ -0,0 +1,502 @@
+                        }
+                    }
+                    return sum;
+                };
+
+                for (u8 u {}; u < 7; ++u) {
+                    double const cu = u == 0 ? inverse_sqrt_2 : 1;
+                    for (u8 v {}; v < 7; ++v) {
+                        auto const table_index = u * 8 + v;
+
+                        double const cv = v == 0 ? inverse_sqrt_2 : 1;
+
+                        // A.3.3 - FDCT and IDCT
+                        double const fdct = cu * cv * sum_xy(u, v) / 4;
+
+                        // A.3.4 - DCT coefficient quantization
+                        i16 const quantized = round(fdct / table.table[table_index]);
+
+                        result[table_index] = quantized;
+                    }
+                }
+
+                for (u8 i {}; i < result.size(); ++i)
+                    component[i] = result[i];
+            };
+
+            convert_one_component(macroblock.y, m_luminance_quantization_table);
+            convert_one_component(macroblock.cb, m_chrominance_quantization_table);
+            convert_one_component(macroblock.cr, m_chrominance_quantization_table);
+        }
+    }
+
+    ErrorOr<void> write_huffman_stream()
+    {
+        for (auto& macroblock : m_macroblocks) {
+            TRY(encode_dc(dc_luminance_huffman_table, macroblock.y, 0));
+            TRY(encode_ac(ac_luminance_huffman_table, macroblock.y));
+
+            TRY(encode_dc(dc_chrominance_huffman_table, macroblock.cb, 1));
+            TRY(encode_ac(ac_chrominance_huffman_table, macroblock.cb));
+
+            TRY(encode_dc(dc_chrominance_huffman_table, macroblock.cr, 2));
+            TRY(encode_ac(ac_chrominance_huffman_table, macroblock.cr));
+        }
+
+        TRY(m_bit_stream.align_to_byte_boundary(0xFF));
+
+        return {};
+    }
+
+    void set_luminance_quantization_table(QuantizationTable const& table, int quality)
+    {
+        set_quantization_table(m_luminance_quantization_table, table, quality);
+    }
+
+    void set_chrominance_quantization_table(QuantizationTable const& table, int quality)
+    {
+        set_quantization_table(m_chrominance_quantization_table, table, quality);
+    }
+
+    QuantizationTable const& luminance_quantization_table() const
+    {
+        return m_luminance_quantization_table;
+    }
+
+    QuantizationTable const& chrominance_quantization_table() const
+    {
+        return m_chrominance_quantization_table;
+    }
+
+    OutputHuffmanTable dc_luminance_huffman_table;
+    OutputHuffmanTable dc_chrominance_huffman_table;
+
+    OutputHuffmanTable ac_luminance_huffman_table;
+    OutputHuffmanTable ac_chrominance_huffman_table;
+
+private:
+    static void set_quantization_table(QuantizationTable& destination, QuantizationTable const& source, int quality)
+    {
+        // The logic to scale the quantization table to a quality coefficient is taken from libjpeg-turbo.
```

Comment:
Apart from the fact that we don't take code from other projects, this would mean an inclusion/mention of their BSD license right? Where did you take this from?

Reply:
IMO, we shouldn't use a custom scale for quality. Other than confusing the user, I don't see the point.
Which means that we should use the same coefficient for our scaling curve as cjpeg, and it's the only thing I've taken from their code.

If you don't consider that as acceptable, I will remove this part for now and do a custom system in another PR.


https://github.com/libjpeg-turbo/libjpeg-turbo/blob/6b9e3b04008165260a13f77cf235170438d5adf8/jcparam.c#L131-L154
```cpp
GLOBAL(int)
jpeg_quality_scaling(int quality)
/* Convert a user-specified quality rating to a percentage scaling factor
 * for an underlying quantization table, using our recommended scaling curve.
 * The input 'quality' factor should be 0 (terrible) to 100 (very good).
 */
{
  /* Safety limit on quality factor.  Convert 0 to 1 to avoid zero divide. */
  if (quality <= 0) quality = 1;
  if (quality > 100) quality = 100;

  /* The basic table is used as-is (scaling 100) for a quality of 50.
   * Qualities 50..100 are converted to scaling percentage 200 - 2*Q;
   * note that at Q=100 the scaling is 0, which will cause jpeg_add_quant_table
   * to make all the table entries 1 (hence, minimum quantization loss).
   * Qualities 1..50 are converted to scaling percentage 5000/Q.
   */
  if (quality < 50)
    quality = 5000 / quality;
  else
    quality = 200 - quality * 2;

  return quality;
}
```

Reply:
It's mainly the wording I think - "taken from" sounds like "copied from" to me, but that's not the case here. We don't want to take over any code verbatim.

Maybe just mention that the calculation is the same?

Reply:
Sounds the same to me, but I literally copied the coefficients.
Anyway, I've updated the comment.